tinybytes/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    borrow, cmp, fmt, hash,
+    ops::{self, RangeBounds},
+    sync::Arc,
+};
+
+/// Immutable bytes type with zero copy cloning and slicing.
+#[derive(Clone)]
+pub struct Bytes {
+    ptr: *const u8,
+    len: usize,
+    // The `bytes`` field is used to ensure that the underlying bytes are freed when there are no
+    // more references to the `Bytes` object. For static buffers the field is `None`.
+    bytes: Option<Arc<dyn UnderlyingBytes>>,
+}
+
+/// The underlying bytes that the `Bytes` object references.
+pub trait UnderlyingBytes: AsRef<[u8]> + Send + Sync + 'static {}
+
+/// Since the Bytes type is immutable, and UnderlyingBytes is `Send + Sync``, it is safe to share
+/// `Bytes` across threads.
+unsafe impl Send for Bytes {}
+unsafe impl Sync for Bytes {}
+
+impl Bytes {
+    /// Creates empty `Bytes`.
+    #[inline]
+    pub fn empty() -> Self {
+        Self::from_static(b"")
+    }
+
+    /// Creates `Bytes` from a static slice.
+    #[inline]
+    pub fn from_static(value: &'static [u8]) -> Self {
+        let slice: &[u8] = value;
+        Self {
+            ptr: slice.as_ptr(),
+            len: slice.len(),
+            bytes: None,
+        }
+    }
+
+    /// Creates `Bytes` from a slice, by copying.
+    pub fn copy_from_slice(data: &[u8]) -> Self {
+        Self::from_underlying(data.to_vec())
+    }
+
+    /// Returns the length of the `Bytes`.
+    #[inline]
+    pub const fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if the `Bytes` is empty.
+    #[inline]
+    pub const fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Returns a slice of self for the provided range.
+    ///
+    /// This will return a new `Bytes` handle set to the slice, and will not copy the underlying
+    /// data.
+    ///
+    /// This operation is `O(1)`.
+    ///
+    /// # Panics
+    ///
+    /// Slicing will panic if the range does not conform to  `start <= end` and `end <= self.len()`.
```

Comment:
```suggestion
    /// Slicing will panic if the range does not conform to  `start <= end` and `end <= self.len()`.
    ///
    /// # Examples
    ///
    /// ```
    /// use tinybytes::Bytes;
    ///
    /// let bytes = Bytes::copy_from_slice(b"hello world");
    /// let slice = bytes.slice(0..5);
    /// assert_eq!(slice.as_ref(), b"hello");
    ///
    /// let slice = bytes.slice(6..11);
    /// assert_eq!(slice.as_ref(), b"world");
    /// ```    
```